Feature FFmpeg plugin

Hrushi20 opened the pull request:
This PR adds FFmpeg Functionality to WasmEdge as a Plugin.

[WasmEdge Rust SDK](https://github.com/Hrushi20/rust-ffmpeg) uses the C++ plugin to provide FFmpeg Functionalities to Wasm Applications.

juntao commented:
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

hydai reviewed (changes requested):
There are too many lines that do not follow the coding style. Please fix them first.

Hrushi20 commented:
Hey @hydai I've followed the coding guidelines as per https://github.com/WasmEdge/WasmEdge/blob/master/.clang-tidy

hydai commented:
Have you run clang-format before?
Also, the DCO check failed. Please fix them.

Hrushi20 commented:
Hey, I've fixed the DCO issues. How do I run the Clang-format file? I'm using the CLion IDE.

hydai commented:
> Hey, I've fixed the DCO issues. How do I run the Clang-format file? I'm using the CLion IDE.

https://www.jetbrains.com/help/clion/clangformat-as-alternative-formatter.html#clion-support


And you have to add your plugin to the workflows to build your plugin in CI. You can refer to this PR for more information: https://github.com/WasmEdge/WasmEdge/pull/2535/files#diff-6322c61775e00a19f0b58dc8e98599df3c6149301a40081bc654228ac306d44f

Hrushi20 commented:
I've added the plugin to workflow and refactored the code based on the style guidelines.

hydai commented:
<img width="529" alt="CleanShot 2023-10-06 at 09 43 01@2x" src="https://github.com/WasmEdge/WasmEdge/assets/2776756/c07c1f17-34a5-49ba-b240-b52a089cc5a5">

There are some conflicts that need to be fixed. Please rebase your commits to the latest master branch.

hydai commented:
Your commits should not contain others' commits. Please fix it.

Hrushi20 commented:
I've squashed all the commits.

hydai commented:
This is incorrect. You have to put your own commits with the latest master. Not using pull/merge.

As you can see in the following picture, your squashed commit still contains some files that are not written by you.
<img width="199" alt="CleanShot 2023-10-08 at 21 27 02@2x" src="https://github.com/WasmEdge/WasmEdge/assets/2776756/88ef00fe-5286-4c64-99d1-b76112b4af77">

Hrushi20 commented:
Can you guide me with the steps?

hydai commented:
1. Checkout a new branch from the latest master
2. Add your commits based on such branch
3. Please always use `git rebase` instead of `git pull` or `git merge` in your local branch, which will produce a useless merge node.
4. When you run the `git rebase`, you may see the conflicts. Fix them, `git add <those files>` and `git rebase --continue`.

Hrushi20 commented:
Hey @hydai, I do I get the CI/CD Pipelines to work. The tests are running successfully on my local pc.

hydai commented:
> Hey @hydai, how do I get the CI/CD Pipelines to work. The tests are running successfully on my local pc.

Your CI are failed. Please make sure you have installed the related dependencies.

Hrushi20 commented:
Hey, I do I ensure deprecated functions are compiled in FFmpeg Plugin? I am using few legacy FFmpeg functions for backward compatibility.

Hrushi20 commented:
Hey @hydai,

Do I split each function into it's own test or follow the current pattern of testing all the functions inside a gigantic GTest function.

https://github.com/Hrushi20/WasmEdge/blob/feature-ffmpeg-plugin/test/plugins/wasmedge_ffmpeg/avutil/avFrame.cpp#L40

hydai commented:
> Hey @hydai,
> 
> Do I split each function into it's own test or follow the current pattern of testing all the functions for a module inside a gigantic GTest function.
> 
> https://github.com/Hrushi20/WasmEdge/blob/feature-ffmpeg-plugin/test/plugins/wasmedge_ffmpeg/avutil/avFrame.cpp#L40

It depends on your cases, normally, we prefer to have a standalone test for each function unless it will need to have a monolithic test.

Hrushi20 commented:
How Do I catch Null pointer exceptions in C++ Plugins??? 

``` c++
Expect<int32_t> SwsScale::body(const Runtime::CallingFrame &, uint32_t SwsCtxId,
                               uint32_t InputFrameId, int32_t SrcSliceY,
                               int32_t SrcSliceH, uint32_t OutputFrameId) {

  FFMPEG_PTR_FETCH(SwsCtx, SwsCtxId, SwsContext);
  FFMPEG_PTR_FETCH(InputFrame, InputFrameId, AVFrame);
  FFMPEG_PTR_FETCH(OutputFrame, OutputFrameId, AVFrame);
  return sws_scale(SwsCtx, InputFrame->data, InputFrame->linesize, SrcSliceY,
                   SrcSliceH, OutputFrame->data, OutputFrame->linesize);
}
```

Assume from Rust SDK, if 0 is passed to InputFrameId, then the Resulting Frame is NULL. Then If I try accessing any fields of the InputFrame, it will lead to NULL pointer exception. 

If I return an error code, then for each function in Rust SDK, I need to validate and check the Error Code.
Is there any way of doing this???

Hrushi20 commented:
Hey @hydai , can you help me with the above question? I'm almost done with adding FFmpeg functionality for version v6. 

I had another approach in mind where we assume that a 0 (Null) is not passed from Rust SDK relying on the Rust Type safe system.

hydai commented:
> How Do I catch Null pointer exceptions in C++ Plugins???
> 
> ```c++
> Expect<int32_t> SwsScale::body(const Runtime::CallingFrame &, uint32_t SwsCtxId,
>                                uint32_t InputFrameId, int32_t SrcSliceY,
>                                int32_t SrcSliceH, uint32_t OutputFrameId) {
> 
>   FFMPEG_PTR_FETCH(SwsCtx, SwsCtxId, SwsContext);
>   FFMPEG_PTR_FETCH(InputFrame, InputFrameId, AVFrame);
>   FFMPEG_PTR_FETCH(OutputFrame, OutputFrameId, AVFrame);
>   return sws_scale(SwsCtx, InputFrame->data, InputFrame->linesize, SrcSliceY,
>                    SrcSliceH, OutputFrame->data, OutputFrame->linesize);
> }
> ```
> 
> Assume from Rust SDK, if 0 is passed to InputFrameId, then the Resulting InputFrame is NULL. Then If I try accessing any fields of the InputFrame, it will lead to NULL pointer exception (SegFault).
> 

You have the `FFMPEG_PTR_FETCH`, what's the issue that you cannot check if the `InputFrame` is `nullptr` or not in the same place?

> If I return an error code, then for each function in Rust SDK, I need to validate and check the Error Code. Is there any way of doing this???

If the `NULL(0)` cannot be passed from Rust, then what's the issue? Just make sure there is no way to pass the `NULL` from the Rust side. Or do you have no way to check the value before passing it?

Hrushi20 commented:
Thanks. I'm ensuring that Null isn't passed from the Rusk Sdk.

Hrushi20 commented:
Hey, can you approve the awaiting workflows?

Hrushi20 commented:
Hey, should the workflows be approved for every git push?

Hrushi20 commented:
Hey, how do I pass values to Test functions? I get the following error while passing values into the function. However, the tests run successfully on my local machine.

<img width="972" alt="Screenshot 2023-11-23 at 1 46 00 PM" src="https://github.com/WasmEdge/WasmEdge/assets/65367566/0c0ef391-dcb2-4844-bf92-b8d12204efa1">

<br><br>
Also, the build of MacOs is failing due to `Permission denied error: mkdir: /usr/local/lib: Permission denied`

hydai commented on plugins/CMakeLists.txt:
```diff
@@ -1,76 +1,80 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: 2019-2022 Second State INC
 
-if(WASMEDGE_PLUGIN_WASI_NN_BACKEND)
```
The indent size should be `2 spaces`.

hydai commented:
1. You should not install ffmpeg by a standalone script. Instead, use `brew install ffmpeg` on macOS.
2. The plugins test failed to build on both manylinux2014(CentOS 7) and ubuntu. What's your environment?

Hrushi20 commented:
> 1. You should not install ffmpeg by a standalone script. Instead, use `brew install ffmpeg` on macOS.
> 
>     2. The plugins test failed to build on both manylinux2014(CentOS 7) and ubuntu. What's your environment?

Currently, the support is only for FFmpeg v6. I'll install FFmpeg v6 using homebrew. 

I am using the Mac M1 environment and building using XCode.

hydai commented:
I also found that your script to build `ffmpeg` is trying to `make install` in a global environment, which is dangerous for users. No one expected installing a dependency on the local build to pollute its environment. Some users are not permitted to install this on their machine; you just encountered the same issue on the macOS CI.
Could you build and install it in a local path?

Hrushi20 commented:
The build for MacOs13 ran successfully 😁. I've used homebrew instead of the build script. 

Do I use apt to install ffmpeg on linux machines?

hydai commented:
> The build for MacOs13 ran successfully 😁. I've used homebrew instead of the build script.
> 
> Do I use apt to install ffmpeg on linux machines?

I don't think so. Because the apt may use a pretty old version of ffmpeg. We can keep the install script temporarily until we migrate this into our CI docker image.

Hrushi20 commented:
I'll keep the bash script. I'll install ffmpeg in local environment. 

However, the tests aren't compiling on linux machines (Based on CI). Is this due to compiler issue on linux?

hydai commented:
> I'll keep the bash script. I'll install ffmpeg in local environment.
> 
> However, the tests aren't compiling on linux machines (Based on CI). Is this due to compiler issue on linux?

I don't think so. There may be something wrong with the include path or cmakefiles. Because the behavior between AppleClang, GCC, and Clang may not be the same. Need to look into it. If you want to debug, use docker for getting the Ubuntu image.

hydai commented:
FYI. I modified the install script for a local usage:

```bash
 #!/usr/bin/env bash

+rm -rf FFmpeg-n6.0 ffmpeg.zip
+
 curl -sL https://github.com/FFmpeg/FFmpeg/archive/refs/tags/n6.0.zip -o ffmpeg.zip

 unzip ffmpeg.zip

 cd FFmpeg-n6.0
-./configure --prefix=/usr/local --enable-gpl --enable-nonfree --enable-shared --disable-static
+mkdir -p output
+./configure --prefix=$(pwd)/output --enable-gpl --enable-nonfree --enable-shared --disable-static
```

You may need to set the `PKG_CONFIG_PATH` for looking for the folder:

```bash
# Under WasmEdge folder
PKG_CONFIG_PATH=$(pwd)/WasmEdge/FFmpeg-n6.0/output/lib/pkgconfig cmake -Bbuild_pr2885 -GNinja -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_BUILD_TESTS=ON  -DWASMEDGE_PLUGIN_FFMPEG=ON .
```

Hrushi20 commented:
> FYI. I modified the install script for a local usage:
> 
> ```shell
>  #!/usr/bin/env bash
> 
> +rm -rf FFmpeg-n6.0 ffmpeg.zip
> +
>  curl -sL https://github.com/FFmpeg/FFmpeg/archive/refs/tags/n6.0.zip -o ffmpeg.zip
> 
>  unzip ffmpeg.zip
> 
>  cd FFmpeg-n6.0
> -./configure --prefix=/usr/local --enable-gpl --enable-nonfree --enable-shared --disable-static
> +mkdir -p output
> +./configure --prefix=$(pwd)/output --enable-gpl --enable-nonfree --enable-shared --disable-static
> ```
> 
> You may need to set the `PKG_CONFIG_PATH` for looking for the folder:
> 
> ```shell
> # Under WasmEdge folder
> PKG_CONFIG_PATH=$(pwd)/WasmEdge/FFmpeg-n6.0/output/lib/pkgconfig cmake -Bbuild_pr2885 -GNinja -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_BUILD_TESTS=ON  -DWASMEDGE_PLUGIN_FFMPEG=ON .
> ```

I've updated the script. Still the build fails due to AOT error.

hydai commented on utils/ffmpeg/install-ffmpeg-v6.0.sh:
```diff
@@ -0,0 +1,17 @@
+echo $(pwd)
+
+#Setting PKG_CONFIG_PATH to be discoverd
+PKG_CONFIG_PATH=$(pwd)/WasmEdge/FFmpeg-n6.0/output/lib/pkgconfig cmake -Bbuild_pr2885 -GNinja -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_BUILD_TESTS=ON  -DWASMEDGE_PLUGIN_FFMPEG=ON .
```
You should not put this line here. This script should install the ffmpeg only.

hydai replied:
The step below the workflow will disable the AOT-related part since there is no need to build it when trying to build the plugins.
Ref: https://github.com/WasmEdge/WasmEdge/pull/2885/files#diff-6322c61775e00a19f0b58dc8e98599df3c6149301a40081bc654228ac306d44fL194

Hrushi20 replied:
Don't we want to compile ffmpeg? By running `make && make install` in the script file, we build ffmpeg in corresponding folder. On compilation we get this `$(pwd)/WasmEdge/FFmpeg-n6.0/output/lib/pkgconfig` path. Else the folder isn't present.

> You should not put this line here. This script should install the ffmpeg only.

Do I need to add this line in github workflow to detect ffmpeg path?

hydai replied:
The `make install` already builds the ffmpeg.
This `PKG_CONFIG_PATH=$(pwd)/WasmEdge/FFmpeg-n6.0/output/lib/pkgconfig ` is for building the ffmpeg plugin. You should add it to the workflow instead of the install script.

Hrushi20 replied:
I've updated the workflow as stated. FFmpeg compiles and build using bash script. However, the FFmpeg Plugin fails to build. It isn't detecting the FFmpeg path. I added it in the workflow.

hydai commented on .github/workflows/build-extensions.yml:
```diff
@@ -185,8 +185,11 @@ jobs:
           # Running tests of wasm_bpf requires proper ebpf running environment
           mount -t debugfs none /sys/kernel/debug
+          bash utils/ffmpeg/install-ffmpeg-v6.0.sh
+          PKG_CONFIG_PATH=$(pwd)/WasmEdge/FFmpeg-n6.0/output/lib/pkgconfig
```
The current folder is `WasmEdge`. Instead, you should use `export PKG_CONFIG_PATH=$(pwd)/FFmpeg-n6.0/output/lib/pkgconfig`.
This line should move to line 195 before you run the cmake command. GitHub action will not keep any environment variables between jobs.

hydai commented on .github/workflows/build-extensions.yml:
```diff
@@ -392,11 +400,14 @@ jobs:
           mount -t debugfs none /sys/kernel/debug
+          # For running ffmpeg tests.
+          bash utils/ffmpeg/install-ffmpeg-v6.0.sh
+          PKG_CONFIG_PATH=$(pwd)/WasmEdge/FFmpeg-n6.0/output/lib/pkgconfig
```
Ditto.

hydai commented on .github/workflows/release.yml:
```diff
@@ -226,6 +226,8 @@ jobs:
           apt install -y libssl-dev cmake g++ wget unzip
           apt install -y libelf-dev zlib1g-dev pkg-config
           apt install -y cargo
+          bash utils/opencvmini/install-opencvmini.sh
```
Why do you add the opencvmini here? This is not your business.

hydai commented on .github/workflows/release.yml:
```diff
@@ -445,6 +453,9 @@ jobs:
           curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | bash -s -- -y
           source "$HOME/.cargo/env"
           bash ./utils/wasi-crypto/build-openssl.sh
+          bash utils/opencvmini/install-opencvmini.sh
```
Ditto.

hydai commented on .github/workflows/release.yml:
```diff
@@ -572,6 +589,7 @@ jobs:
         run: |
           eval $(/opt/homebrew/bin/brew shellenv)
           brew install llvm@16 ninja cmake openssl wabt opencv rust gh
+          bash utils/ffmpeg/install-ffmpeg-v6.0.sh
```
The same thing, you should always use `brew install` on macOS.

hydai commented on plugins/CMakeLists.txt:
```diff
@@ -1,76 +1,81 @@
 # SPDX-FileCopyrightText: 2019-2022 Second State INC
 
-if(WASMEDGE_PLUGIN_WASI_NN_BACKEND)
+if (WASMEDGE_PLUGIN_WASI_NN_BACKEND)
```
You should never change the coding style here. There is no space between `if` and `(`.

Hrushi20 commented:
FFmpeg is being installed locally and that is being used to run the CI tests 😄. The tests in MacOs is being built, however in Linux I get `undefined reference error`. Why is that happening?

hydai commented:
> FFmpeg is being installed locally and that is being used to run the CI tests 😄.
> 
> The tests in MacOs is being built, however in Linux I get `undefined reference error` while testing the Functions defined in FFmpeg Plugin. Why is that happening?

I still believe there is something wrong with the included path or cmakefile. The reason is that there are many other plugins that are built without any issues. It cannot be a problem with the environment or the building steps.

Hrushi20 commented:
How do proceed with Merging the PR. I know it's a very long Pull Request. Any suggestions on making it simpler. 

I'm still looking into why the plugin fails on Linux machines. I'm not able to figure out the issue.

Hrushi20 commented:
> > FFmpeg is being installed locally and that is being used to run the CI tests 😄.
> > The tests in MacOs is being built, however in Linux I get `undefined reference error` while testing the Functions defined in FFmpeg Plugin. Why is that happening?
> 
> I still believe there is something wrong with the included path or cmakefile. The reason is that there are many other plugins that are built without any issues. It cannot be a problem with the environment or the building steps.

Could the reason be that the test files have the same name and folder structure as the Plugin files?

hydai commented:
> How do proceed with Merging the PR? I know it's a very long Pull Request. Any suggestions on making it simpler.
1. You may use squash to reduce the commit numbers.
2. Once you are ready, say the CI is all green, then I can take more time to review it. Please let me know if you get ready.
> I'm still looking into why the plugin fails on Linux machines. I'm not able to figure out the issue.
1. Since the other plugins pass the CI, so the environment and toolchain should be fine.
2. One possible issue is that the path(include path) or some configuration in CMake is wrong. Please verify them.

hydai commented:
> Could the reason be that the test files have the same folder and naming structure as the Plugin files?

Maybe, I am not digging into this issue since I am working on the other PR. Please do some experiments on it.

Hrushi20 commented:
I'm currently wrapping the tests on the Rust SDK. Will look into this in 2-3 days.

Hrushi20 commented:
I installed a ubuntu machine and tried debugging the issue. I found that the build is successful when I am compiling with the Clang compiler. If I am using the gcc compiler it fails. 

I am not sure what kind of issue is this.

Hrushi20 commented:
I installed clang just for testing purpose on my ubuntu machine. The code built successfully. 
If I set the below flag, then the tests run successfully. 
```console
$ export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
```

Without the above flag I get the following error:

<img width="1237" alt="Screenshot 2023-12-07 at 8 39 13 PM" src="https://github.com/WasmEdge/WasmEdge/assets/65367566/67de4e7f-eb00-4702-aed2-dff49ce84d70">

Hrushi20 commented:
I have gone through the CMakeCache files on Ubuntu.

Below is CMakeCache file without clang. The code fails to build (CI)
<img width="533" alt="Screenshot 2023-12-10 at 12 20 28 PM" src="https://github.com/WasmEdge/WasmEdge/assets/65367566/e1b8b4d6-e9b5-4903-a166-156f0286ccc7">
 Ubuntu.

Below is CMakeCache file with CLang. It uses LLVM behind the scene. This code compiles and runs successfully.
<img width="533" alt="Screenshot 2023-12-10 at 12 25 14 PM" src="https://github.com/WasmEdge/WasmEdge/assets/65367566/fd8a89bf-41f2-4cf9-9311-e499579d949a">

hydai commented:
> I have gone through the CMakeCache files on Ubuntu.
So, the current status is as follows:
GCC -> Failed
Clang -> Fine

It sounds like there is an undefined behavior or something that is not the same between GCC and Clang.

Hrushi20 commented:
> > I have gone through the CMakeCache files on Ubuntu.
> > So, the current status is as follows:
> > GCC -> Failed
> > Clang -> Fine
> 
> It sounds like there is an undefined behavior or something that is not the same between GCC and Clang.

The source files are compiling as expected. It's only the test files that are not detecting the source files.

Hrushi20 commented:
I got the linux build to run successfully. It was an issue in CMake file. Using `wasmedge_add_library` fails to build on linux. If I use 'add_library', the test code successfully complies. 

I am not sure why the windows tests are failing? They are independent of the plugin.

hydai commented:
And for the Windows one, do you add the gmock for the windows target? It looks like the gmock on Windows is not setting up correctly.

Hrushi20 commented:
> I also found your code is not passed in the spell-check:
> 
> Please use valid names instead. If you want to keep this, please add several exceptions in [ignore_word](https://github.com/WasmEdge/WasmEdge/blob/master/.github/workflows/ignore_words)
> 
> ```
> plugins/wasmedge_ffmpeg/avfilter/avFilter.cpp:76: Inout ==> Input, In out
> plugins/wasmedge_ffmpeg/avfilter/avFilter.cpp:91: Inout ==> Input, In out
> plugins/wasmedge_ffmpeg/avfilter/avFilter.cpp:101: Inout ==> Input, In out
> plugins/wasmedge_ffmpeg/avfilter/avFilter.cpp:110: Inout ==> Input, In out
> plugins/wasmedge_ffmpeg/avfilter/avfilter_func.cpp:91: InOuts ==> inputs
> plugins/wasmedge_ffmpeg/avfilter/avfilter_func.cpp:92: InOuts ==> inputs
> plugins/wasmedge_ffmpeg/avfilter/avfilter_func.cpp:194: Inout ==> Input, In out
> plugins/wasmedge_ffmpeg/avfilter/avfilter_func.cpp:[19](https://github.com/WasmEdge/WasmEdge/actions/runs/7164787328/job/19505493132?pr=2885#step:4:20)5: Inout ==> Input, In out
> plugins/wasmedge_ffmpeg/avfilter/avfilter_func.cpp:196: Inout ==> Input, In out
> plugins/wasmedge_ffmpeg/avfilter/avfilter_func.cpp:198: Inout ==> Input, In out
> test/plugins/wasmedge_ffmpeg/avfilter/avfilter_func.cpp:54: anull ==> annul
> ```

I added the words to `ignore_words` file. The linter still fails.

Hrushi20 commented:
I named all the functions Inout and the fields as InOut. It's better to have InOut to maintain consistency.

Hrushi20 commented:
```
/usr/bin/ld: _deps/spdlog-build/libspdlog.a: error adding symbols: file format not recognized
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

The CI fails on Ubuntu using CLang compiler. What does the above error mean?

Hrushi20 commented:
Hey @hydai, the above code is ready for review.

hydai reviewed (changes requested):
This is a huge PR, I still need time to finish the review, the current progress is 49 / 108 files viewed.

Here are some common issues:
1. Use `nullptr`. Don't `NULL`.
2. Use `static_cast<>()`. Don't `(type)var`.
3. Remove redundant blank lines.
4. Make sure the variables are fit the naming styles.

Hrushi20 commented:
Hey, any update regarding the PR?

hydai commented:
@Hrushi20 
Could you please rebase to the latest master? We are reducing the workflow files to make the building process more clear.

I am going to review your PR next week. Sorry for the delay.

Hrushi20 commented:
Is there any change in env? The clang compiler fails on Ubuntu and g++ compiler runs successfully. I get the below error in CI.

```
-- Checking for modules 'libavdevice;libavfilter;libavformat;libavcodec;libswresample;libswscale;libavutil'
--   No package 'libavdevice' found
--   No package 'libavfilter' found
--   No package 'libavformat' found
--   No package 'libavcodec' found
--   No package 'libswresample' found
--   No package 'libswscale' found
--   No package 'libavutil' found
```

hydai commented:
We only changed the way of building the WASI-NN plugin: https://github.com/WasmEdge/WasmEdge/pull/2986/files#diff-6322c61775e00a19f0b58dc8e98599df3c6149301a40081bc654228ac306d44fR285-R290
The remaining plugins should not be affected.

Hrushi20 commented:
Hey @hydai, I've fixed the build. The test fails for MacOs on CI but they build on my local Mac machine. Any idea what can be happening?

hydai commented:
> Hey @hydai, I've fixed the build. The test fails for MacOs on CI but they build on my local Mac machine. Any idea what can be happening?

It could be many reasons, such as the OS version, the toolchain version, the third-party dependencies version, and any related part. That's why we have to make sure the CI is green, which is a reproducible environment.

Hrushi20 commented:
Hey @hydai, my windows build is failing. This is due to gtest throwing an error. I feel this is causing an issue in macos build as well.

hydai commented:
It looks like the include path has something wrong: `D:\a\WasmEdge\WasmEdge\build\_deps\gtest-src\googlemock\include\gmock/internal/gmock-port.h(56): fatal error C1083: Cannot open include file: 'gtest/internal/gtest-port.h': No such file or directory`

Hrushi20 commented:
@hydai can you reveiw the code? I've fixed the failing CI on macos. I was passing a wrong type to ffmpeg function in test.

Hrushi20 commented:
I've updated the test suite as well :)

hydai commented:
@Hrushi20 
The plugin CI failed. Could you please rebase to the latest master?
It seems like something is going wrong after the workflow is modified.

Hrushi20 commented:
The tests are all running. There is an issue with zipping the plugins, due to which the pipeline is failing.

Hrushi20 commented:
@hydai, I've rebased the code. Prepare the WasmEdge plugin phase fails (manylinux). I've compared the workflow.yml files with origin/master and there is no difference except addition of ffmpeg plugin.

hydai commented on .github/workflows/build-extensions.yml:
```diff
@@ -321,19 +330,19 @@ jobs:
+      build_options_all_platforms: -DWASMEDGE_PLUGIN_WASI_CRYPTO=ON -DWASMEDGE_PLUGIN_WASI_LOGGING=ON -DWASMEDGE_PLUGIN_PROCESS=ON -DWASMEDGE_PLUGIN_TENSORFLOW=ON -DWASMEDGE_PLUGIN_TENSORFLOWLITE=ON -DWASMEDGE_PLUGIN_IMAGE=ON -DWASMEDGE_PLUGIN_OPENCVMINI=ON -DWASMEDGE_PLUGIN_FFMPEG=ON
       build_options_manylinux2014_x86_64: -DWASMEDGE_PLUGIN_WASM_BPF=ON -DWASMEDGE_PLUGIN_WASM_BPF_BUILD_LIBBPF_WITH_PKG_CONF=OFF
       build_options_manylinux2014_aarch64:
-      tar_names_all_platforms: wasi_crypto wasi_logging wasmedge_process wasmedge_tensorflow wasmedge_tensorflowlite wasmedge_image wasmedge_opencvmini
+      tar_names_all_platforms: wasi_crypto wasi_logging wasmedge_process wasmedge_tensorflow wasmedge_tensorflowlite wasmedge_image wasmedge_opencvmini wasmedge_ffmpeg
       tar_names_manylinux2014_x86_64: wasm_bpf
       tar_names_manylinux2014_aarch64:
-      test_bins_all_platforms: wasiCryptoTests wasiLoggingTests wasmedgeProcessTests wasmedgeTensorflowTests wasmedgeTensorflowLiteTests wasmedgeImageTests wasmedgeOpencvminiTests
+      test_bins_all_platforms: wasiCryptoTests wasiLoggingTests wasmedgeProcessTests wasmedgeTensorflowTests wasmedgeTensorflowLiteTests wasmedgeImageTests wasmedgeOpencvminiTests wasmedgeFFmpegTests
       test_bins_manylinux2014_x86_64: wasmBpfTests
       test_bins_manylinux2014_aarch64:
-      output_bins_all_platforms: libwasmedgePluginWasiCrypto.so libwasmedgePluginWasiLogging.so libwasmedgePluginWasmEdgeProcess.so libwasmedgePluginWasmEdgeTensorflow.so libwasmedgePluginWasmEdgeTensorflowLite.so libwasmedgePluginWasmEdgeImage.so libwasmedgePluginWasmEdgeOpenCVMini.so
+      output_bins_all_platforms: libwasmedgePlugintatasiCrypto.so libwasmedgePluginWasiLogging.so libwasmedgePluginWasmEdgeProcess.so libwasmedgePluginWasmEdgeTensorflow.so libwasmedgePluginWasmEdgeTensorflowLite.so libwasmedgePluginWasmEdgeImage.so libwasmedgePluginWasmEdgeOpenCVMini.so libwasmedgePluginWasmEdgeFFmpeg.so
```
You modified the name of `libwasmedgePluginWasiCrypto`. That's why it broke the CI.

hydai commented on plugins/wasmedge_ffmpeg/avformat/avformatContext.cpp:
```diff
@@ -0,0 +1,124 @@
+Expect<uint32_t> AVFormatCtxNbStreams::body(const Runtime::CallingFrame &,
+                                            uint32_t AvFormatCtxId) {
+
+  auto ffmpegMemory = Env.get();
```
Naming style.

hydai commented on plugins/wasmedge_ffmpeg/avformat/avformatContext.cpp:
```diff
@@ -0,0 +1,124 @@
+                                            uint32_t AvFormatCtxId) {
+
+  auto ffmpegMemory = Env.get();
+  AVFormatContext *avFormatCtx =
```
Ditto.

hydai commented on plugins/wasmedge_ffmpeg/avutil/avDictionary.cpp:
```diff
@@ -0,0 +1,159 @@
+
+  FFMPEG_PTR_FETCH(SrcAvDict, SrcDictId, AVDictionary *);
+
+  int Res;
```
You must initialize the value.

hydai commented on plugins/wasmedge_ffmpeg/avutil/pixfmt.cpp:
```diff
@@ -0,0 +1,199 @@
+namespace WasmEdgeFFmpeg {
+namespace AVUtil {
+
+// Expect<int32_t> AVPixFmtDescGet::body(const Runtime::CallingFrame
```
Why does this section comment out? If we don't need it, remove it. Otherwise, it would be best if you kept it unimplemented or something meaningful.

hydai commented on plugins/wasmedge_ffmpeg/avutil/pixfmt.cpp:
```diff
@@ -0,0 +1,199 @@
+  return static_cast<int32_t>(PixelFormat);
+}
+
+// Expect<int32_t> AvPixFmtDescriptorName::body(const Runtime::CallingFrame
```
Ditto.

hydai commented on plugins/wasmedge_ffmpeg/avutil/pixfmt.h:
```diff
@@ -0,0 +1,148 @@
+namespace WasmEdgeFFmpeg {
+namespace AVUtil {
+//
+// class AVPixFmtDescGet : public WasmEdgeFFmpegAVUtil<AVPixFmtDescGet> {
```
Ditto

hydai commented on plugins/wasmedge_ffmpeg/avutil/pixfmt.h:
```diff
@@ -0,0 +1,148 @@
+                       uint32_t PixFormatId);
+};
+//
+// class AvPixFmtDescriptorName : public
```
Ditto.

hydai commented on test/plugins/wasmedge_ffmpeg/avcodec/avCodec.cpp:
```diff
@@ -0,0 +1,365 @@
+      dynamic_cast<WasmEdge::Host::WasmEdgeFFmpeg::AVcodec::AVCodecID &>(
+          FuncInst->getHostFunc());
+
+  spdlog::info("Testing AVCodecId");
```
Don't forget to append `sv` when using the string literals.

hydai reviewed (changes requested):
Almost ready to get this merge. Here are some minor issues that should be fixed:

1. Please double-check the naming style; some variables are in the wrong format.
2. Please double-check the initialization of a using variable.
3. Please append `sv` when using the string literals.

hydai commented on test/plugins/wasmedge_ffmpeg/swscale/swscale_func.cpp:
```diff
@@ -0,0 +1,545 @@
+      FuncInst->getHostFunc());
+
+  {
+
```
I found that there are so many useless empty lines; please remove all of them.

hydai replied:
Adding an empty line after `{` is nonsense; could you explain why these lines are needed?

hydai commented on test/plugins/wasmedge_ffmpeg/swscale/swscale_func.cpp:
```diff
@@ -0,0 +1,545 @@
+          FuncInst->getHostFunc());
+
+  {
+    uint32_t swsVecId = readUInt32(MemInst, SwsVectorPtr);
```
The same issue is that several variables are in the wrong naming style.

hydai reviewed (changes requested):
Minor issues need to be fixed.

Hrushi20 commented:
Can we release the plugin in WasmEdge version `0.14.0`?

hydai commented on utils/ffmpeg/download-ffmpeg-sample-video.sh:
```diff
@@ -0,0 +1,20 @@
+# SPDX-FileCopyrightText: 2019-2022 Second State INC
+
+TODIR=$1
+SAMPLE_VIDEO=https://raw.githubusercontent.com/Hrushi20/rust-ffmpeg/master/assets/bunny.mp4
```
I found that this issue is not resolved. Because using an unknown resource will have the license issue. This is a blocker to get this PR merged.

hydai commented on utils/ffmpeg/download-ffmpeg-sample-video.sh:
```diff
@@ -0,0 +1,20 @@
+
+if [ ! -f $TODIR/sample_video.mp4 ]; then
+  curl -sL $SAMPLE_VIDEO -o $TODIR/sample_video.mp4
+  cp $TODIR/sample_video.mp4 $TODIR/dummy.mp4 # Dummy file to manipulate and run tests on file.
```
Also, this script downloads the sample video, but there is no way to delete it.

hydai commented on test/plugins/wasmedge_ffmpeg/avformat/avformat_func.cpp:
```diff
@@ -0,0 +1,548 @@
+    EXPECT_TRUE(Result[0].get<int32_t>() >= 0);
+  }
+
+  //  FuncInst =
```
This is not answered.

hydai commented on test/plugins/wasmedge_ffmpeg/avfilter/avfilter_func.cpp:
```diff
@@ -0,0 +1,682 @@
+    EXPECT_EQ(Result[0].get<int32_t>(), static_cast<int32_t>(ErrNo::Success));
+  }
+
+  // Crashing the program. Checked even from Rust side.
```
Is this a bug that needs to be fixed?

hydai commented on test/plugins/wasmedge_ffmpeg/avcodec/avcodec_func.cpp:
```diff
@@ -0,0 +1,575 @@
+TEST_F(FFmpegTest, AVCodecFunc) {
+  ASSERT_TRUE(AVCodecMod != nullptr);
+
+  //  std::string FileName = "ffmpeg-assets/sample_video.mp4"; // 32 chars
```
If this is commented out, why not delete it?

hydai reviewed (changes requested):
Some questions need to be answered. We will merge this PR when all concern issues are fixed.

hydai commented:
> Can we release the plugin in WasmEdge version `0.14.0`?

Depends on the remaining issues. If everything gets resolved, then we can have this in 0.14.0

Hrushi20 commented:
```
I found that this issue is not resolved. Because using an unknown resource will have the license issue. This is a blocker to get this PR merged.
```

I've taken the above video from [ffmpeg-libav-tutorial](https://github.com/leandromoreira/ffmpeg-libav-tutorial/blob/master/LICENSE) repository. The repository contains required license.

Hrushi20 commented:
```
Also, this script downloads the sample video, but there is no way to delete it.
```
I don't see the wasi-nn models being deleted in the tests as well. How do I delete the sample video after executing the tests?

hydai commented:
> ```
> I found that this issue is not resolved. Because using an unknown resource will have the license issue. This is a blocker to get this PR merged.
> ```
> 
> I've taken the above video from [ffmpeg-libav-tutorial](https://github.com/leandromoreira/ffmpeg-libav-tutorial/blob/master/LICENSE) repository. The repository contains required license.

You should add a credit for it and show the link. This is very important to us.

hydai commented:
> ```
> Also, this script downloads the sample video, but there is no way to delete it.
> ```
> 
> I don't see the wasi-nn models being deleted in the tests as well. How do I delete the sample video after executing the tests?

After checking the codes, the only way to remove these test assets is to remove the whole test folder. So it's fine then.

Hrushi20 commented:
> > ```
> > I found that this issue is not resolved. Because using an unknown resource will have the license issue. This is a blocker to get this PR merged.
> > ```
> > 
> >     
> >   
> > I've taken the above video from [ffmpeg-libav-tutorial](https://github.com/leandromoreira/ffmpeg-libav-tutorial/blob/master/LICENSE) repository. The repository contains required license.
> 
> You should add a credit for it and show the link. This is very important to us.

I've added the credit and link into the bash script file.

hydai commented:
Thanks.
